docs/advanced-usage.md

```diff
@@ -99,25 +112,80 @@ Now, only certain interaction methods are allowed to be called at certain points
+Thus, if `.isEnding()` is `true`, the only valid method you may call is `.start()` to restart the conversation session from the beginning.
 
-### Response Methods
+## Events
 
-A **response method** is any method of `RuntimeClient` that looks like this: `.on___(callbackFn)` (i.e. `.onSpeak`). The `callbackFn` will look like this: `(trace, context) => ...` These get triggered whenever an interaction method is finished. Response methods allow you control the behavior of your app whenever based on the blocks in your Voiceflow project - by iterating through the `context.getTrace()` and matching the correct trace type.
+The `RuntimeClient` has an event system that notifies the developer of any changes in the `RuntimeClient`'s data. 
 
-To see what the trace contains, try something like `onSpeak((trace) => console.log(trace))`. For the definition of traces, see [Advanced Trace Types](#advanced-trace-types)
+### Event Types
+
+Trace Events when the `RuntimeClient` receives a response from our Runtime servers. For **all** traces that `RuntimeClient` receives from our Runtime servers - not just the traces you've configured to see with the `includeTypes` option - we trigger a corresponding event for that trace.
+
+The full list of events is listed below.
+
+- `TraceType.X` - When a specific trace of type `X` is being processed, there is a corresponding event that is fired, e.g., if `SpeakTrace` is received then the `TraceType.SPEAK` event is triggered.
+- `TRACE_EVENT` - Triggered when any trace is being processed.
+
+Moreover, Trace Events are **guaranteed to occur in the order of the trace response**. For example, if the `RuntimeClient` received a list containing `BlockTrace`, `SpeakTrace`, `DebugTrace`, `SpeakTrace` in that order, then the following events will occur in this exact order:
+
+- `TraceType.BLOCK`
+- `TRACE_EVENT`
+- `TraceType.SPEAK` - Corresponds with the first `SpeakTrace` in the list
+- `TRACE_EVENT`
+- `TraceType.DEBUG`
+- `TRACE_EVENT`
+- `TraceType.SPEAK` - Corresponds with the second `SpeakTrace` in the list
+- `TRACE_EVENT`
+
+### Event Handlers
+
+To register an event handler, use the below methods:
+
+- `.on(event: TraceType | TRACE_EVENT, handler: Function)` - This method is used to register `handler` on the given `event`
+- `.onSpeak(handler: Function)` - This method is used to register `handler` on a `TraceType.SPEAK` event. There exists equivalents for all other trace types as well.
+
+Note, since Trace Events occur in the order of the trace response, then handlers also execute sequentially in the order. That is, we call handlers for a trace in the response list, only after all previous traces in the list are handled.
+
+```ts
+rclient.on(TraceType.SPEAK, (trace, context) => {		// register a handler for only SpeakTraces
+  myStore.traces.push(trace.payload.message);				// traces will be added to your local store in order
```

Comment:
not sure why you would be adding the trace to the local store? you can just do `console.log(trace.payload.message)`

Reply:
Done